File: heap/230.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int[] res = new int[]{0};
        inorder(root, k, new int[]{1}, res);
        return res[0];
    }

    public void inorder(TreeNode root, int k, int[] count, int[] res) {
        if (root == null) {
            return;
        }

        inorder(root.left, k, count, res);

        if (count[0] == k){
            res[0] = root.val;
        }
        count[0] = count[0] + 1;

        inorder(root.right, k, count, res);
    }
}

File: problems/310/medium/_py/230.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        res = []
        
        def inOrderDfs(root):
            if root is None:
                return False

            if inOrderDfs(root.left):
                return True

            res.append(root.val)
            if len(res) == k:
                return True

            if inOrderDfs(root.right):
                return True
        
        inOrderDfs(root)
        return res[-1]
